File: src/main/scala/vm/vagrant/configuration/builder/VagrantEnvironmentConfigBuilder.scala
package vm.vagrant.configuration.builder

/**
  * Created by oliver.ziegert on 24.03.17.
  */

import java.io.File

import scala.collection.immutable.List
import vm.vagrant.configuration.VagrantEnvironmentConfig
import vm.vagrant.configuration.VagrantVmConfig
import vm.vagrant.configuration.builder.util.VagrantBuilderException


object VagrantEnvironmentConfigBuilder {
  def create = new VagrantEnvironmentConfigBuilder
}

class VagrantEnvironmentConfigBuilder() {
  private var vmConfigs: List[VagrantVmConfig] = List()
  private var path: File = _
  private var version: String = _

  def withVagrantVmConfig(vmConfig: VagrantVmConfig): VagrantEnvironmentConfigBuilder = {
    vmConfigs ::= vmConfig
    this
  }

  def withPath(path: File): VagrantEnvironmentConfigBuilder = {
    this.path = path
    this
  }

  def withVersion(version: String): VagrantEnvironmentConfigBuilder = {
    this.version = version
    this
  }

  def build: VagrantEnvironmentConfig = {
    if (vmConfigs.isEmpty) throw new VagrantBuilderException("No vm defined")
    if (path == null) throw new VagrantBuilderException("No path defined")
    if (version == null || version.isEmpty) new VagrantBuilderException("No version defined")
    new VagrantEnvironmentConfig(vmConfigs = vmConfigs,
                                 path = path,
                                 version = version)
  }
}

File: src/main/java/vm/vagrant/configuration/builder/VagrantEnvironmentConfigBuilder.java
package vm.vagrant.configuration.builder;

import vm.vagrant.configuration.VagrantEnvironmentConfig;
import vm.vagrant.configuration.VagrantVmConfig;
import vm.vagrant.configuration.builder.util.VagrantBuilderException;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class VagrantEnvironmentConfigBuilder {

    private List<VagrantVmConfig> vmConfigs;
    private File path;

    public VagrantEnvironmentConfigBuilder() {
        vmConfigs = new ArrayList<>();
    }

    public static VagrantEnvironmentConfigBuilder create() {
        return new VagrantEnvironmentConfigBuilder();
    }

    public VagrantEnvironmentConfigBuilder withVagrantVmConfig(
            VagrantVmConfig vmConfig) {
        this.vmConfigs.add(vmConfig);
        return this;
    }

    public VagrantEnvironmentConfigBuilder withPath(File path) {
        this.path = path;
        return this;
    }

    public VagrantEnvironmentConfig build() {
        if (vmConfigs.isEmpty()) {
            throw new VagrantBuilderException("No vm defined");
        }
        return new VagrantEnvironmentConfig(vmConfigs, path);
    }
}
